storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,574 @@
+// Obvious con is that stale reads might occur which we are fine with in this scenario
+func loadAnnounces(etcd *clientv3.Client, timeout time.Duration) ([]memory.Announce, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	rangeResp, err := etcd.Get(ctx, EtcdAnnouncePrefix, clientv3.WithPrefix(), clientv3.WithSerializable())
+	if err != nil {
+		return nil, fmt.Errorf("failed to do etcd operation, bootstrap of all infohash-peer announces: %v", err)
+	}
+
+	announces := make([]memory.Announce, 0, rangeResp.Count)
+	log.Info(fmt.Sprintf("existing announces found = %d", rangeResp.Count))
+	for _, kv := range rangeResp.Kvs {
+		ih, peerKey, seeder, err := decodeAnnounceKey(kv.Key)
+		if err != nil {
+			log.Error(fmt.Sprintf("error decoding announce key %s: %v", kv.Key, err))
+			continue
+		}
+		clock, err := strconv.ParseInt(string(kv.Value), 10, 64)
+		if err != nil {
+			log.Error(fmt.Sprintf("Invalid clock value read for key-value %s-%s: %v", kv.Key, kv.Value, err))
+			continue
+		}
+		announces = append(announces, memory.NewAnnounce(ih, peerKey, seeder, clock))
+	}
+
+	return announces, nil
+}
+
+type serializedPeer []byte
+
+func newPeerKey(p *bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+	return b
+}
+
+const (
+	EtcdKeySep                = "/"
+	EtcdAnnouncePrefix        = "announce"
+	EtcdAnnounceSeederPrefix  = "s" //len 1
+	EtcdAnnounceLeecherPrefix = "l" //len 1
+)
+
+// Format of announce key in etcd:
+// <announce_prefix><sep><infohash><sep><peer_type><sep><peer_key>
+// peer key as returned by newPeerKey and uniquely identifies a peer in swarm
+// peer type can either be s(seeder) or l(leecher)
+func newAnnounceKey(ih *bittorrent.InfoHash, p *bittorrent.Peer, seeder bool) []byte {
+	var peerType []byte
+	if seeder {
+		peerType = []byte(EtcdAnnounceSeederPrefix)
+	} else {
+		peerType = []byte(EtcdAnnounceLeecherPrefix)
+	}
+
+	var pk serializedPeer
+	if p != nil {
+		pk = newPeerKey(p)
+	}
+
+	key := bytes.Join([][]byte{
+		[]byte(EtcdAnnouncePrefix),
+		ih[:],
+		peerType,
+		pk,
+	}, []byte(EtcdKeySep))
+
+	return key
+}
+
+func decodeAnnounceKey(key []byte) (*bittorrent.InfoHash, []byte, bool, error) {
+	announceKeyReader := func() func() ([]byte, error) {
+		sep := []byte(EtcdKeySep)
+		partLengths := []int{len([]byte(EtcdAnnouncePrefix)), 20, 1}
+		begin, end, currPart := 0, 0, 0
+		return func() ([]byte, error) {
+			if currPart == len(partLengths)+1 {
+				//all parts of announce key have been read
+				return nil, nil
+			}
+			if currPart < len(partLengths) {
+				end = begin + partLengths[currPart]
+				if end > len(key) {
+					return nil, fmt.Errorf("invalid key, length:%d shorter than expected:%d", len(key), end)
+				}
+				read := key[begin:end]
+				begin = end + len(sep)
```

Comment:
Better to check it the `key[end:end+len(sep)] == sep` ?

Reply:
Good point, will fix